server/src/main/java/org/elasticsearch/snapshots/RestoreService.java

```diff
@@ -723,19 +722,18 @@ public static RestoreInProgress updateRestoreStateWithDeletedIndices(RestoreInPr
         boolean changesMade = false;
         RestoreInProgress.Builder builder = new RestoreInProgress.Builder();
         for (RestoreInProgress.Entry entry : oldRestore) {
-            ImmutableOpenMap.Builder<ShardId, ShardRestoreStatus> shardsBuilder = null;
+            Map<ShardId, ShardRestoreStatus> shards = null;
             for (Map.Entry<ShardId, ShardRestoreStatus> cursor : entry.shards().entrySet()) {
                 ShardId shardId = cursor.getKey();
                 if (deletedIndices.contains(shardId.getIndex())) {
                     changesMade = true;
-                    if (shardsBuilder == null) {
-                        shardsBuilder = ImmutableOpenMap.builder(entry.shards());
+                    if (shards == null) {
+                        shards = new HashMap<>(entry.shards());
                     }
-                    shardsBuilder.put(shardId, new ShardRestoreStatus(null, RestoreInProgress.State.FAILURE, "index was deleted"));
+                    shards.put(shardId, new ShardRestoreStatus(null, RestoreInProgress.State.FAILURE, "index was deleted"));
                 }
             }
-            if (shardsBuilder != null) {
-                ImmutableOpenMap<ShardId, ShardRestoreStatus> shards = shardsBuilder.build();
+            if (shards != null) {
```

Comment:
I think we need to wrap the mutable HashMap in Collections.unmodifiableMap if it's not null down below when we create a new RestoreInProgress.Entry.

Reply:
`Entry` is a public class, so the constructor should be extra-careful about immutableness in the constructor, just wrapping `Collections.unmodifiableMap` leaves the possibility for the caller to modify the passed-in collection afterwards